docs/sat-solver.md

```diff
@@ -0,0 +1,383 @@
+kind of problem is an optimization problem in
+<img src="https://render.githubusercontent.com/render/math?math=f: \{0,1\}^n \mapsto \Re">,
+Over _n_ Boolean variables
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}, \dots ,x_{n}">.
+In our case, whether package _x_ is present (installed) in the final solution.
+The optimization problem form is in the following function to minimize:
+
+<img src="https://render.githubusercontent.com/render/math?math=min \sum_{j \in N} c_{j}*x_{j}">
+
+subject to:
+
+<img src="https://render.githubusercontent.com/render/math?math=\sum_{j \in N} a_{ij}*l_{j} \geq b_{i}">
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{j} \in \{0,1\}, a_{ij},b_{i}, c_{j} \in N_{0}^{+}, i \in M">
+
+<img src="https://render.githubusercontent.com/render/math?math=M = 1, \dots, m">
+
+Where:
+
+- Each <img src="https://render.githubusercontent.com/render/math?math=c_{j}">
+  is a non-negative integer cost, associated with variable
+  <img src="https://render.githubusercontent.com/render/math?math=x_{j}, j \in N">.
+- <img src="https://render.githubusercontent.com/render/math?math=a_{ij}">
+  denotes the coefficients of the literals
+  <img src="https://render.githubusercontent.com/render/math?math=l_{j}">
+  in the set m of linear constraints (a literal being a Boolean variable).
+
+In our case, a package _x_ belongs to the repository _P_, where _P_ is the set
+of tuples
+<img src="https://render.githubusercontent.com/render/math?math=P = \{ (repoUrl, name, semverRange, ns), \dots \}">.
+
+## Constraints definitions
+
+We will define Pseudo-Boolean functions that will be optimized by selecting
+their input values. This Pseudo-Boolean functions can then be implemented in
+Hypper easily.
+
+### Installation
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}"> is installed:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} \geq 1">,
+with <img src="https://render.githubusercontent.com/render/math?math=x_{1} \in P">
+
+For simplicity, the tuple of package
+<img src="https://render.githubusercontent.com/render/math?math=x = (repoUrl, name, semverRange = 1.0, ns)">
+is represented as
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">.
+
+
+### Depends
+
+<img src="https://render.githubusercontent.com/render/math?math=p_{1}"> depends on
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} - p_{1} \geq 0">
+
+Installing <img src="https://render.githubusercontent.com/render/math?math=p_{1}"> implies
+installing <img src="https://render.githubusercontent.com/render/math?math=x_{1}">.
+Although <img src="https://render.githubusercontent.com/render/math?math=x_{1}"> may be
+installed without <img src="https://render.githubusercontent.com/render/math?math=p_{1}">.
+
+
+### Depends on multiple versions
+
```

Comment:
Not clear what "multiple versions" means here.

Reply:
True. Maybe better wording is "Depends, that can be satisfied by multiple versions".

Reply:
reworded in https://github.com/rancher-sandbox/hypper/pull/171/commits/7b7120ebe2cfb7f08def8d16dc5f10457c1ff2f8.

Reply:
reworded in https://github.com/rancher-sandbox/hypper/pull/171/commits/7b7120ebe2cfb7f08def8d16dc5f10457c1ff2f8.